posts/building-a-realtime-app-with-phoenix.md

```diff
@@ -0,0 +1,604 @@
+2. Teach our client how to initialize a WebSocket connection
+3. Teach our client how to join a specific channel
+4. Teach our client how to broadcast a message across that channel
+
+Let's get started.
+
+### Step 1: Declare a WebSocket Endpoint
+
+We need to do two things to get our WebSocket endpoint up and running:
+
+* Declare the socket endpoint
+* Route requests to that endpoint to the correct channel
+
+#### Defining the Endpoint
+
+When you generate a new Phoenix app, your app's endpoint gets generated for you for free. 
+
+The [application endpoint](https://hexdocs.pm/phoenix/Phoenix.Endpoint.html) acts as the starting point for all requests to your web app. This is where we'll define our socket endpoint.
+
+```elixir
+defmodule PhoenixPair.Endpoint do
+  use Phoenix.Endpoint, otp_app: :phoenix_pair
+
+  socket "/socket", PhoenixPair.UserSocket
+end
+```
+
+This line is where the magic happens:
+
+```elixir
+socket "/socket", PhoenixPair.UserSocket
+```
+
+Here, we are declaring a socket on the `/socket` URI and telling our app that the `UserSocket` module will handle all such connection requests. 
+
+
+#### Routing Socket Requests to Channels
+
+Our app generation also gave us the `UserSocket` module for free in `/channels/user_socket.ex`.
+
+The `UserSocket` is responsible for making sure our socket requests get routed to the correct channel:
+
+```elixir
+defmodule PhoenixPair.UserSocket do
+  use Phoenix.Socket
+
+  channel "challenges:*", PhoenixPair.ChallengeChannel
+end
+```
+
+Whenever a client sends a message whose topic starts with "challenges:", it will be routed to our `ChallengeChannel`. 
+
+Now that our socket endpoint is up and running, let's tell our client to open a socket connection.
+
+### Step 2: Connecting a Client to the Socket
+
+We'll have our client send a socket connect request to the socket endpoint. Then, the endpoint will forward that request to `UserSocket`, which will establish the connection. 
+
+![](../images/real-time-phoenix-image-one.svg)
+
+#### Sending the Connect Request From the Client
+
+Phoenix Channels provides a JavaScript library that we can use to send a socket connection request from the client.
+
+
+```javascript
+import { Socket } from 'phoenix';
+const socket = new Socket("/socket", {params: {token: <your auth token>}})
+socket.connect();
+```
+
+Calling `socket.connect()` on the front-end sends a request to our `/socket` endpoint, which routes the request to our `UserSocket` module.
+
+#### Handling the Connect Request on the Server
+
+The `UserSocket` module receives the connect request and invokes the `connect` function with an argument of the params that we passed to our `Socket` initialization. 
+
+
+This `connect` function is where we perform some authorization actions. 
+
+```elixir
+# user_socket.ex
+defmodule PhoenixPair.UserSocket do
+  use Phoenix.Socket
+  alias PhoenixPair.{GuardianSerializer}
```

Comment:
Don't think you need the multi import here, should be `alias PhoenixPair.GuardianSerializer`.